src/Evaluator.hs

```diff
@@ -50,313 +450,221 @@ evalErrorText (EvalErrorCase loc val) =
-  pure $ ValueInteger i
-eval (Core (CoreSyntax syntax)) = do
-  pure $ ValueSyntax syntax
-eval (Core (CoreCase loc scrutinee cases)) = do
-  v <- eval scrutinee
-  doCase loc v cases
-eval (Core (CoreIdent (ScopedIdent ident scope))) = do
-  identSyntax <- evalAsSyntax ident
-  case identSyntax of
-    Syntax (Stx _ _ expr) ->
-      case expr of
-        Integer _ ->
-          throwError $ EvalErrorType $ TypeError
-            { _typeErrorExpected = "id"
-            , _typeErrorActual   = "integer"
-            }
-        String _ ->
-          throwError $ EvalErrorType $ TypeError
-            { _typeErrorExpected = "id"
-            , _typeErrorActual   = "string"
-            }
-        List _ ->
-          throwError $ EvalErrorType $ TypeError
-            { _typeErrorExpected = "id"
-            , _typeErrorActual   = "list"
-            }
-        Id name -> withScopeOf scope $ Id name
-eval (Core (CoreEmpty (ScopedEmpty scope))) = withScopeOf scope (List [])
-eval (Core (CoreCons (ScopedCons hd tl scope))) = do
-  hdSyntax <- evalAsSyntax hd
-  tlSyntax <- evalAsSyntax tl
-  case tlSyntax of
-    Syntax (Stx _ _ expr) ->
-      case expr of
-        List vs -> withScopeOf scope $ List $ hdSyntax : vs
-        String _ ->
-          throwError $ EvalErrorType $ TypeError
-            { _typeErrorExpected = "list"
-            , _typeErrorActual   = "string"
-            }
-        Id _ ->
-          throwError $ EvalErrorType $ TypeError
-            { _typeErrorExpected = "list"
-            , _typeErrorActual   = "id"
-            }
-        Integer _ ->
-          throwError $ EvalErrorType $ TypeError
-            { _typeErrorExpected = "list"
-            , _typeErrorActual   = "integer"
-            }
-eval (Core (CoreList (ScopedList elements scope))) = do
-  vec <- List <$> traverse evalAsSyntax elements
-  withScopeOf scope vec
-eval (Core (CoreIntegerSyntax (ScopedInteger int scope))) = do
-  intV <- evalAsInteger int
-  withScopeOf scope (Integer intV)
-eval (Core (CoreStringSyntax (ScopedString str scope))) = do
-  strV <- evalAsString str
-  withScopeOf scope (String strV)
-eval (Core (CoreReplaceLoc loc stx)) = do
-  Syntax (Stx _ newLoc _) <- evalAsSyntax loc
-  Syntax (Stx scs _ contents) <- evalAsSyntax stx
-  return $ ValueSyntax $ Syntax $ Stx scs newLoc contents
-eval (Core (CoreTypeCase loc scrut cases)) = do
-  ty <- evalAsType scrut
-  env <- ask
-  return $ ValueMacroAction $ MacroActionTypeCase env loc ty cases
-
-evalErrorType :: Text -> Value -> Eval a
+  in evaluateIn env _closureBody
+apply (HO prim) value = return $! prim value
+
+applyAsClosure :: VEnv -> Value -> Value -> Kont -> EState
+applyAsClosure e v_closure value k = case v_closure of
+    ValueClosure closure -> app closure
+    other                -> Er (evalErrorType "function" other) e k
+
+    where app (FO (FOClosure{..})) =
+            let env = Env.insert _closureVar _closureIdent value (_closureEnv <> e)
```

Comment:
This is a bug. When applying a closure, the body of the closure must be evaluated in the captured environment, not in the union of the captured environment plus the environment at the call site. That is, the following code should not be accepted because `y` is not in scope where `f` is defined:
```
f = \() -> y
g y = f ()
```

:x: this is a blocker.

Reply:
I'm having a hard time writing a test case for this because the patch errors as expected with (I believe) the correct error message. Here is my translation to klister:

```
#lang kernel

(define f (lambda (ff) y))
(define g (lambda (y) (f)))

(example (g 2))
```

Note that I have to pass `ff` or else I get and error about no arguments:

```
❯ cat examples/non-examples/bad-lexical-env.golden
Expected 1 entries between parentheses, but got
  #[bad-lexical-env.kl:3.19-3.22]<()>
```

this is a sidenote but I'm surprised we cannot define a function of type `f : () -> <whatever>`.

Here is what main shows:

```
Test suite klister-tests: RUNNING...
All tests
  Golden tests
    bad-lexical-env: OK (0.01s)
...
❯ cat examples/non-examples/bad-lexical-env.golden
Unknown: #[bad-lexical-env.kl:4.24-4.25]<y>
```

Notice that this is what we should expect, an error saying `unknown y`.

and here is the patch:

```
❯ cabal test --test-show-details=streaming --test-options '-p bad-lexical-env --accept'
...
Test suite klister-tests: RUNNING...
All tests
  Golden tests
    bad-lexical-env: OK (0.01s)
      Accepted the new version
...
❯ cat examples/non-examples/bad-lexical-env.golden
Unknown: #[bad-lexical-env.kl:3.24-3.25]<y>
```

That is, I accept the new output for the golden test and we get the same error as `main` from the patch.

Reply:
> I'm having a hard time writing a test case for this

Makes sense: scope errors are caught at expansion time, precisely so that they cannot occur at evaluation-time, so it should not be possible to write a Klister program which triggers this bug. Okay, this isn't as big of an issue as I thought then.

✏️ this is no longer a blocker, but after this MR is finally merged, I plan to come back and fix this.

Reply:
Note that I have to pass ff or else I get and error about no arguments:

```
❯ cat examples/non-examples/bad-lexical-env.golden
Expected 1 entries between parentheses, but got
  #[bad-lexical-env.kl:3.19-3.22]<()>
```

this is a sidenote but I'm surprised we cannot define a function of type `f : () -> <whatever>`.

I agree that it is surprising, but it makes sense if you think about it. `(lambda () 42)` is a function from _zero_ arguments, not a function from the unit type to Integer. Racket supports functions from zero arguments, but Haskell does not. Like Haskell, Klister only supports functions of one argument, so `(lambda (x y) 42)` desugars to `(lambda (x) (lambda (y) 42))` and if it wasn't rejected `(lambda () 42)` would desugar to `42`. The error message could definitely be improved though!

Reply:
great explanation! You're right, I guess I see some scheme or lisp and assume that I can define a function with 0-arity.